Integration test for embeded Android views touch support.

amirh opened the pull request:
*This depends on #19907, #20052, and flutter/goldens#10, devicelab configuration will be added in a following PR.*

This test verifies that the synthesized motion events that get to embedded
Android view are equal to the motion events that originally hit the FlutterView.

The test app's Android code listens to MotionEvents that get to FlutterView and
to an embedded Android view and sends them over a platform channel to the Dart
code, where the events are matched.

This is what the app looks like:

![android_views test app](https://flutter.github.io/assets-for-api-docs/assets/readme-assets/android_views_test.png)

The blue part is the embedded Android view, because it is positioned at the top
left corner, the coordinate systems for FlutterView and for the embedded view's
virtual display has the same origin (this makes the MotionEvent comparison
easier as we don't need to translate the coordinates).

The app includes the following control buttons:
  * RECORD - Start listening for MotionEvents for 3 seconds, matched/unmatched events are
    displayed in the listview as they arrive.
  * CLEAR - Clears the events that were recorded so far.
  * SAVE - Saves the events that hit FlutterView to a file.
  * PLAY FILE - Send a list of events from a bundled asset file to FlutterView.

A recorded touch events sequence is bundled as an asset in the
assets_for_android_view package which lives in the goldens repository.

When running this test with `flutter drive` the record touch sequences is
replayed and the test asserts that the events that got to FlutterView are
equivalent to the ones that got to the embedded view.

amirh committed:
Integration test for embeded Android Views touch support.

The test places an embedded Android view at the top left, and verifies
that motion events that get to FlutterView are equivalent to the
synthesized motion events that gets to the embedded view.

See the README.md for more high level details.

xster commented on dev/integration_tests/android_views/README.md:
```diff
@@ -0,0 +1,32 @@
+# Integration test for touch events on embedded Android views
```
Are you adding more generic Android view embedding tests to this or is it just for touch events? (just wondering if you're going to want to use this generic name later)

xster commented on dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java:
```diff
@@ -0,0 +1,106 @@
+    mMethodChannel = new MethodChannel(this.getFlutterView(), "android_views_integration");
+    mMethodChannel.setMethodCallHandler(this);
+    mFlutterViewTouchPipe = new TouchPipe(mMethodChannel, getFlutterView());
+
```
extra line

xster commented on dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java:
```diff
@@ -0,0 +1,106 @@
+import io.flutter.plugins.GeneratedPluginRegistrant;
+
+public class MainActivity extends FlutterActivity implements MethodChannel.MethodCallHandler {
+
```
uber nit: be consistent with extra line at top and bottom of scopes

xster commented on dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java:
```diff
@@ -0,0 +1,126 @@
+
+@TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
+public class MotionEventCodec {
+
```
ditto

xster commented on dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java:
```diff
@@ -0,0 +1,126 @@
+public class MotionEventCodec {
+
+    public static HashMap<String, Object> encode(MotionEvent event) {
+        ArrayList<HashMap<String,Object>> pointerProperties = new ArrayList<>();
```
be consistent with block indents

amirh replied:
done

xster commented on dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java:
```diff
@@ -0,0 +1,106 @@
+        getExternalStoragePermissions();
+        return;
+      case "synthesizeEvent":
+          synthesizeEvent(methodCall, result);
```
Keep consistent spacing

xster commented on dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java:
```diff
@@ -0,0 +1,126 @@
+        List<PointerProperties> pointerProperties = new ArrayList<>();
+        List<PointerCoords> pointerCoords = new ArrayList<>();
+
+        for (HashMap<String, Object> property : (List<HashMap<String, Object>>)data.get("pointerProperties")) {
```
https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace `#5`, add a space after casting or be consistent in the file.

amirh replied:
done

xster commented on dev/integration_tests/android_views/lib/main.dart:
```diff
@@ -0,0 +1,297 @@
+            height: 300.0,
+            child: new AndroidView(
+                viewType: 'simple_view',
+                onPlatformViewCreated: onPlatformViewCreated)),
```
indents and trailing commas

amirh replied:
done

xster commented on dev/integration_tests/android_views/pubspec.yaml:
```diff
@@ -0,0 +1,22 @@
+  path_provider: ^0.4.1
+  collection: ^1.14.6
+  assets_for_android_views:
+    path: ../../../bin/cache/pkg/goldens/dev/integration_tests/assets_for_android_views
```
reminder to edit this before submitting

amirh replied:
Clarified f2f - we don't publish this to pub, we're actually fetching the goldens in the driver test and reading from this path.

xster commented on dev/integration_tests/android_views/lib/motion_event_diff.dart:
```diff
@@ -0,0 +1,191 @@
+
+String diffMotionEvents(
+    Map<String, dynamic> originalEvent,
+    Map<String, dynamic> synthesizedEvent) {
```
trailing commas

amirh replied:
done

xster commented on dev/integration_tests/android_views/lib/motion_event_diff.dart:
```diff
@@ -0,0 +1,191 @@
+  diffMaps(expected, actual, diffBuffer,
+      messagePrefix: '[pointerCoord $pointerIdx] ',
+      excludeKeys: <String>[
+        'size', // Currently the framework doesn't get the size form the engine.
```
*from

amirh replied:
done

xster commented on dev/integration_tests/android_views/lib/motion_event_diff.dart:
```diff
@@ -0,0 +1,191 @@
+
+void diffMaps(Map<String, dynamic> expected,
+    Map<String, dynamic> actual, StringBuffer diffBuffer,
+{List<String> excludeKeys = const <String>[],
```
some indent issues. one per line

amirh replied:
done

xster commented on dev/integration_tests/android_views/lib/main.dart:
```diff
@@ -0,0 +1,297 @@
+      children: <Widget>[
+        new SizedBox(
+            height: 300.0,
+            child: new AndroidView(
```
Feel free to ignore or do in a different PR: I wonder if we can integrate a 'performance test' in here. Besides the synthesized down time and event times, can we wrap this whole thing in a transparent gesture detector, record the actual time here and put that into a list as well and also record the time on the native size, pass it back along with the motionevents and make sure the time delta doesn't get past a threshold?

amirh replied:
I plan to add a benchmark for the roundtrip of MotionEvent->PointerEvents->MotionEvent using this integration test.

xster reviewed (commented):
LGTM

amirh committed:
flutter update-packages

amirh committed:
Clone goldens as part of flutter update-packages.

flutter_tools cannot depend on flutter_goldens (as flutter_goldens
depdends on the Flutter sdk), so this commit splits client.dart from
flutter_goldens to a pure-dart flutter_goldens_client package.

amirh committed:
Canonicalize path depdendencies in `flutter update-packages`

The tool was failing if we had 2 separate non-canonical paths to the
same canonical paths.

amirh commented:
@tvolkert I split a flutter_goldens_client packages out of flutter_goldens.
Could you take a look just at: 876f907 and 99f8a95

(I'll audit the commit history into a sequence that makes more sense before landing)

tvolkert reviewed (approved):
LGTM

fkorotkov commented:
@Hixie I think they hit an edge case and the agent stuck in a never ending retry loop on downloading a cache. It will soon timeout and the tasks will be automatically re-run. I will work on a fix for the retry issue. Sorry about that and thanks for letting me know.

fkorotkov commented:
@Hixie the infinite retry thing should be fixed now. 🙌
